File: apps/ml/ui/src/types/annotator.ts
export interface AnnotatorProfile {
  email: string
  full_name: string
  phone_number: string | null
  country: string
  county: string
  bio: string
  total_points_earned: number
  total_points_redeemed: number
  redeemable_points: number
  total_entries_submitted: number
  total_tasks_completed: number
  referral_code: string
  joined_at: string
  last_active_at: string | null
  // KYC
  kyc_status: 'none' | 'pending' | 'approved' | 'rejected'
  avatar_key: string | null
  id_front_key: string | null
  id_back_key: string | null
  kyc_submitted_at: string | null
  kyc_rejection_reason: string | null
}

export interface AnnotatorStats {
  total_entries: number
  total_points: number
  redeemable_points: number
  total_tasks: number
  tasks_completed: number
  kes_value: number          // legacy: local amount
  local_amount: number
  local_currency: string
  local_formatted: string
  rank: number | null
}

export interface AnnotatorTask {
  dataset_id: string
  name: string
  description: string
  category: string
  points_enabled: boolean
  points_per_entry: number
  points_redemption_info: string
  require_location?: boolean
  field_count?: number
  entry_count: number
  total_entries: number
  joined: boolean
  token: string | null
  is_repeatable?: boolean
  is_done?: boolean
  required_fields_count?: number
  status?: string
}

export interface RewardSummary {
  total_earned: number
  total_redeemed: number
  redeemable: number
  kes_value: number         // legacy
  local_amount: number
  local_currency: string
  local_formatted: string
  rate_label: string
  min_redemption_points: number
  can_redeem: boolean
  kyc_required: boolean
  kyc_status: string
}

export interface RewardRedemption {
  id: string
  points_redeemed: number
  kes_value: number
  phone_number: string
  status: 'pending' | 'sent' | 'failed'
  created_at: string
}

export interface PlatformRewardRate {
  point_value_usd: number
  currency: string
  one_point_value: string
  hundred_points_value: string
  rate_label: string
  exchange_rates: Record<string, number>
  min_redemption_points: number
  withdrawal_kyc_threshold_usd: number
}

File: apps/ml/app/models/annotator.py
"""Platform-level annotator profiles and reward redemptions."""
import secrets
from typing import Optional
from datetime import datetime
from beanie import Document
from pymongo import IndexModel, ASCENDING
from pydantic import Field
from app.utils.datetime import utc_now


class AnnotatorProfile(Document):
    """One document per annotator user. Platform-scoped (no org_id)."""

    class Settings:
        name = "annotator_profiles"
        indexes = [IndexModel([("email", ASCENDING)], unique=True)]

    email: str
    full_name: str = ""
    phone_number: Optional[str] = None      # for airtime redemption
    country: str = "KE"                     # KE | TZ | UG | RW | ET | NG etc.
    county: str = ""            # county/state within country
    bio: str = ""
    total_points_earned: int = 0
    total_points_redeemed: int = 0
    redeemable_points: int = 0              # earned - redeemed
    total_entries_submitted: int = 0
    total_tasks_completed: int = 0
    referral_code: str = Field(default_factory=lambda: secrets.token_urlsafe(8))
    referred_by: Optional[str] = None       # referral_code of the referrer
    joined_at: datetime = Field(default_factory=utc_now)
    last_active_at: Optional[datetime] = None
    # KYC
    kyc_status: str = "none"   # none | pending | approved | rejected
    avatar_key: Optional[str] = None
    id_front_key: Optional[str] = None
    id_back_key: Optional[str] = None
    kyc_submitted_at: Optional[datetime] = None
    kyc_reviewed_at: Optional[datetime] = None
    kyc_rejection_reason: Optional[str] = None


class RewardRedemption(Document):
    """Airtime redemption record."""

    class Settings:
        name = "reward_redemptions"

    annotator_email: str
    points_redeemed: int
    kes_value: float                        # KES value sent
    phone_number: str
    at_transaction_id: Optional[str] = None  # Africa's Talking transaction ID
    status: str = "pending"                  # pending | sent | failed
    error_message: Optional[str] = None
    created_at: datetime = Field(default_factory=utc_now)
    updated_at: datetime = Field(default_factory=utc_now)
